Model Message, Schema and RecordBatch metadata in the NIF

Benjamin-Philip opened the pull request:
This commit adds support for modelling the Message, Schema, and RecordBatch metadata (plural).
Additionally support for encoding and decoding messages, schemas and record batches as well as tests have been added.

Benjamin-Philip committed:
Add Test

This commit adds a test for the stub `print` function to print a Message
from Rust. This function tests if a message record is decoded correctly
into Rust.

Benjamin-Philip committed:
Add Support for Decoding Schema Messages

This commit adds support for decoding just schema messages.

Benjamin-Philip committed:
Add Support for Decoding Record Batches

Benjamin-Philip committed:
Disable clippy::needless_borrow for RecordBatch

Benjamin-Philip commented:
I've never used Rustler (and honestly also Rust) for anything important before. Could you please let me know if my work so far looks good? I've got some commented code to delete, after which we could merge.

I plan to add decoding support later.

Benjamin-Philip committed:
Remove commented code

Benjamin-Philip committed:
Replace print with test_decode

Printing should be avoided in tests, and should be restricted to
debugging. Instead we can just return ok. test_decode is the previous
print function without the printing parts.

Benjamin-Philip committed:
Add the test_encode function and CommonTest test

Benjamin-Philip committed:
Fix Clippy Error

Benjamin-Philip committed:
Support Encoding

This commit adds support for encoding the messages.

Benjamin-Philip commented on native/arrow_format_nif/src/message.rs:
```diff
@@ -0,0 +1,18 @@
+#[derive(Debug, NifRecord)]
+#[tag = "message"]
+pub struct Message {
+    pub version: Atom, // TODO Limit this to just v1..v5.
```
Would [`UnitEnums`](https://docs.rs/rustler/latest/rustler/derive.NifUnitEnum.html) be good for this?

polvalente replied:
Looks like it's exactly what you need :)

My suggestion would be basically what it accomplishes: declare an enum, and derive the encoder and decoder traits to convert to and from elixir atoms

Benjamin-Philip commented on native/arrow_format_nif/src/message/header.rs:
```diff
@@ -0,0 +1,29 @@
+use super::schema::Schema;
+use rustler::{Decoder, Encoder, Term};
+
+#[derive(Debug)]
+pub enum Header {
+    Schema(Schema),
+    RecordBatch(RecordBatch),
+}
```
As well as for this

Benjamin-Philip replied:
Actually, [`NifUntaggedEnum`](https://docs.rs/rustler_codegen/0.29.1/src/rustler_codegen/lib.rs.html#360) worked best with this.

Benjamin-Philip commented on native/arrow_format_nif/src/message/record_batch.rs:
```diff
@@ -0,0 +1,23 @@
+    pub length: i64,
+    pub nodes: Vec<FieldNode>,
+    pub buffers: Vec<Buffer>,
+    pub compression: Atom, // TODO limit this to undefined, lz4_frame, and zstd
```
this,

Benjamin-Philip commented on native/arrow_format_nif/src/message/schema.rs:
```diff
@@ -0,0 +1,15 @@
+#[derive(Debug, NifRecord)]
+#[tag = "schema"]
+pub struct Schema {
+    pub endianness: Atom, // TODO limit this to little and big
```
and this?

Benjamin-Philip reviewed (commented):
I have a few questions on using UnitEnums:

Benjamin-Philip commented on native/arrow_format_nif/src/message/schema/field.rs:
```diff
@@ -0,0 +1,42 @@
+
+// TODO: Update Option<T>'s Encoder and Decoder implementations to treat
+// undefined as None in rustler main.
+#[derive(Debug)]
+pub enum Name {
+    Atom(Atom),
+    String(String),
+}
+
+impl Encoder for Name {
+    fn encode<'a>(&self, env: rustler::env::Env<'a>) -> Term<'a> {
+        match &self {
+            Name::Atom(_) => crate::atoms::undefined().encode(env),
+            Name::String(name) => name.clone().into_bytes().encode(env),
+        }
+    }
+}
+
+impl<'a> Decoder<'a> for Name {
+    fn decode(term: Term<'a>) -> rustler::NifResult<Name> {
+        if term.is_atom() {
+            Ok(Name::Atom(term.decode()?))
+        } else {
+            let vector: Vec<u8> = term.decode()?;
+            let string = String::from_utf8(vector).unwrap();
+            Ok(Name::String(string))
+        }
+    }
+}
```
Does rustler support charlists <--> String decoding and encoding built-in (In which case I can use ErlOption)?

Benjamin-Philip replied:
I'm guessing that it's best to leave this implemented by hand.

polvalente replied:
The code here seems simple enough as it is

Benjamin-Philip committed:
Restrict versions, features, endianness, and compression

This commmit restricts the values possible in the version field in the
Message struct, the features, and endianness fields in the Schema struct
and the compression field in the RecordBatch struct with the use of
UnitEnums.

Benjamin-Philip committed:
Replace hand-written Header with NifUntaggedEnum

This commit replaces the hand-written Encoder and Decoder
implementations for Header with the ones generated with NifUntaggedEnum.

polvalente commented on native/arrow_format_nif/src/lib.rs:
```diff
@@ -1 +1,195 @@
+        v4,
+        v5,
+
+        // Endiannes
```
```suggestion
        // Endianness
```

polvalente commented on native/arrow_format_nif/src/lib.rs:
```diff
@@ -1 +1,195 @@
+                endianness: Endianness::Little,
+                fields: vec![
+                    Field {
+                        name: Name::String("id".to_string()),
```
I wonder if you could implement the `From<&str>` trait so that you could do `Name::from("id")`

polvalente replied:
I believe it would be:

```rust
impl From<&str> for Name {
    fn from(value: &str) -> Self {
        Name::String(value.to_string())
    }
}
```

polvalente commented on native/arrow_format_nif/src/message/record_batch.rs:
```diff
@@ -0,0 +1,23 @@
+#![allow(clippy::needless_borrow)]
```
What's the error you get if you remove it?

Benjamin-Philip replied:
You can see the error in the CI log before "Remove Clippy Error". I'll paste the error here once I'm free.

Benjamin-Philip replied:
```
$ cargo clippy
    Checking arrow_format_nif v0.1.0 (/home/ben/Github/serde_arrow/native/arrow_format_nif)
warning: the borrowed expression implements the required traits
  --> src/message/record_batch.rs:15:5
   |
15 |     pub length: i64,
   |     ^^^^^^^^^^ help: change this to: `pub length`
   |
   = note: `#[warn(clippy::needless_borrow)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

warning: the borrowed expression implements the required traits
  --> src/message/record_batch.rs:16:5
   |
16 |     pub null_count: i64,
   |     ^^^^^^^^^^^^^^ help: change this to: `pub null_count`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

warning: the borrowed expression implements the required traits
  --> src/message/record_batch.rs:21:5
   |
21 |     pub offset: i64,
   |     ^^^^^^^^^^ help: change this to: `pub offset`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

warning: the borrowed expression implements the required traits
  --> src/message/record_batch.rs:22:5
   |
22 |     pub length: i64,
   |     ^^^^^^^^^^ help: change this to: `pub length`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

warning: `arrow_format_nif` (lib) generated 4 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 1.12s
```

Benjamin-Philip replied:
The suggested changes actually don't compile.

polvalente commented on native/arrow_format_nif/src/message/schema/field.rs:
```diff
@@ -0,0 +1,42 @@
+}
+
+// TODO: Update Option<T>'s Encoder and Decoder implementations to treat
+// undefined as None in rustler main.
```
I didn't understand this comment

Benjamin-Philip replied:
This comment was from before I was aware of `ErlOption`. In Erlang, the convention is to use `undefined` instead of `nil`, so `Option` doesn't work. This comment was just a reminder for me to write something for ErlOption to eliminate the Name Enum.

Then I realised that rustler has poor charlist support.

I will delete this comment.

josevalim replied:
> Then I realised that rustler has poor charlist support.

It may be worth moving to binaries for strings anyway. Erlang projects such as Hackney and Cowboy already do that.

Benjamin-Philip replied:
I probably should. But sometime later.

polvalente commented on native/arrow_format_nif/src/message/schema/field.rs:
```diff
@@ -0,0 +1,42 @@
+    fn encode<'a>(&self, env: rustler::env::Env<'a>) -> Term<'a> {
+        match &self {
+            Name::Atom(_) => crate::atoms::undefined().encode(env),
+            Name::String(name) => name.clone().into_bytes().encode(env),
```
Do you need .clone() if you're doing .into_bytes()?

Benjamin-Philip replied:
The code fails to compile without cloning.
I'll paste the error when I'm free.

Benjamin-Philip replied:
```
$ cargo b
   Compiling arrow_format_nif v0.1.0 (/home/ben/Github/serde_arrow/native/arrow_format_nif)
error[E0507]: cannot move out of `*name` which is behind a shared reference
   --> src/message/schema/field.rs:27:35
    |
27  |             Name::String(name) => name.into_bytes().encode(env),
    |                                   ^^^^^------------
    |                                   |    |
    |                                   |    `*name` moved due to this method call
    |                                   move occurs because `*name` has type `String`, which does not implement the `Copy` trait
    |
note: this function takes ownership of the receiver `self`, which moves `*name`
   --> /home/ben/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/string.rs:867:23
    |
867 |     pub fn into_bytes(self) -> Vec<u8> {
    |                       ^^^^

For more information about this error, try `rustc --explain E0507`.
error: could not compile `arrow_format_nif` due to previous error
```

Benjamin-Philip committed:
Update native/arrow_format_nif/src/lib.rs

Co-authored-by: Paulo Valente <16843419+polvalente@users.noreply.github.com>

Benjamin-Philip committed:
Remove comment on Option<T>

Benjamin-Philip committed:
Implement From<&str> for Name

This will make using Name a little more ergonomic.

Benjamin-Philip commented:
Merging as approved by @polvalente .
